nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/ListHDFS.java

```diff
@@ -263,24 +267,35 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
             final FileStatusManager fileStatusManager = new FileStatusManager(latestTimestamp, latestFiles);
             final Path rootPath = getNormalizedPath(context, DIRECTORY);
-            final FileStatusIterable fileStatuses = new FileStatusIterable(rootPath, recursive, hdfs, getUserGroupInformation());
+            final FileStatusIterable fileStatusIterable = new FileStatusIterable(rootPath, recursive, hdfs, getUserGroupInformation());
 
             final Long minAgeProp = context.getProperty(MINIMUM_FILE_AGE).asTimePeriod(TimeUnit.MILLISECONDS);
             final long minimumAge = (minAgeProp == null) ? Long.MIN_VALUE : minAgeProp;
             final Long maxAgeProp = context.getProperty(MAXIMUM_FILE_AGE).asTimePeriod(TimeUnit.MILLISECONDS);
             final long maximumAge = (maxAgeProp == null) ? Long.MAX_VALUE : maxAgeProp;
 
-            final HadoopFileStatusWriter writer;
-            if (writerFactory == null) {
-                writer = new FlowFileObjectWriter(session, fileStatuses, minimumAge, maximumAge, pathFilter, fileStatusManager, latestTimestamp, latestFiles);
-            } else {
-                writer = new RecordObjectWriter(session, fileStatuses, minimumAge, maximumAge, pathFilter, fileStatusManager, latestTimestamp,
-                        latestFiles, writerFactory, getLogger());
-            }
+            final HadoopWriterContext hadoopWriterContext = HadoopWriterContext.builder()
+                    .session(session)
+                    .successRelationship(getSuccessRelationship())
+                    .fileStatusIterable(fileStatusIterable)
+                    .fileStatusManager(fileStatusManager)
+                    .pathFilter(pathFilter)
+                    .minimumAge(minimumAge)
+                    .maximumAge(maximumAge)
+                    .previousLatestTimestamp(latestTimestamp)
+                    .previousLatestFiles(latestFiles)
+                    .writerFactory(writerFactory)
+                    .hdfsPrefix(getAttributePrefix())
+                    .logger(getLogger())
+                    .build();
+
+            final HadoopFileStatusWriter writer = (writerFactory == null)
```

Comment:
Adding a builder looks to be a great step forward! It could be even more beneficial for us if the builder would not only build the context but the writer and the actual writer instance creation would be hidden behind the `build` method. I think, adding an extra method which can help the builder decide what kind of output we expect (maybe a small enum) would help with this and this way also the context and the writer implementations might remain unexposed.